kapitan/refs/secrets/azkms.py

```diff
@@ -0,0 +1,221 @@
+            if encode_base64:
+                kwargs["encoding"] = "base64"
+        else:
+            self.data = data
+        super().__init__(self.data, **kwargs)
+        self.type_name = 'azkms'
+
+
+    @classmethod
+    def from_params(cls, data, ref_params):
+        """
+        Return new Secret encrypted with a Azure KMS key
+        from data and ref_params: target_name
+        key will be grabbed from the inventory via target_name
+        """
+        try:
+            target_name = ref_params.kwargs['target_name']
+            if target_name is None:
+                raise ValueError('target_name not set')
+
+            target_inv = cached.inv['nodes'].get(target_name, None)
+            if target_inv is None:
+                raise ValueError('target_inv not set')
+
+            key = target_inv['parameters']['kapitan']['secrets']['azkms']['key']
+            vault = target_inv['parameters']['kapitan']['secrets']['azkms']['vault']
+            return cls(data, key, vault, **ref_params.kwargs)
+        except KeyError:
+            raise RefError("Could not create AzureKMSSecret: target_name missing")
+
+    @classmethod
+    def from_path(cls, ref_full_path, **kwargs):
+        return super().from_path(ref_full_path, encrypt=False)
+
+    def reveal(self):
+        """
+        returns decrypted data
+        raises AzureKMSError if decryption fails
+        """
+        # can't use super().reveal() as we want bytes
+        ref_data = base64.b64decode(self.data)
+        return self._decrypt(ref_data, self.key, self.vault)
+
+    def update_key(self, key):
+        """
+        re-encrypts data with new key, respects original encoding
+        returns True if key is different and secret is updated, False otherwise
+        """
+        if key == self.key:
+            return False
+
+        data_dec = self.reveal()
+        encode_base64 = self.encoding == 'base64'
+        if encode_base64:
+            data_dec = base64.b64decode(data_dec).decode()
+        self._encrypt(data_dec, key, encode_base64)
+        self.data = base64.b64encode(self.data).decode()
+        return True
+
+    def _encrypt(self, data, key, vault, encode_base64):
+        """
+        encrypts data
+        set encode_base64 to True to base64 encode data before writing
+        """
+        assert isinstance(key, str)
+        _data = data
+        self.encoding = "original"
+        if encode_base64:
+            _data = base64.b64encode(data.encode())
+            self.encoding = "base64"
+        else:
+             # To guarantee _data is bytes
+             if isinstance(data, str):
+                _data = data.encode()
+        try:
+            ciphertext = ""
+            # Mocking encrypted response for tests
+            if key == "mock":
+                ciphertext = base64.b64encode("mock".encode())
+            else:
+            #
+            # TODO: Get the encryption algo, currently hardcoded to EncryptionAlgorithm.rsa_oaep_256
```

Comment:
The algorithm choice here is another thing that could be defined in the inventory along with the key + vault url. :)

Reply:
ok